### ISSUE TITLE
Minimize lodash dependency

### ISSUE THREAD

**Issue by exoego** [serverless/serverless#7747]
### Use case description

https://github.com/serverless/serverless/pull/7680#discussion_r422044932

`lodash` offers lots of utility, but nowadays we only need the small parts of `lodash`, since recent Node.js supports native convenient counterpart (e.g. `Array#forEach(f)` over `_.forEach(array, f)`).

Minimizing `lodash` dependency will achieve

* Faster runtime speed by native implementation
* Smaller bundle size of serverless framework
* Friendly for developers who are more familiar with standard JS, not lodash.

### Proposed solution

* Use convenient native counterpart as much as possible
* Depend only some part of `lodash`, e.g. `lodash-memoize` and so on.

__Note__: Ideally if we handle all below cases, and each is addressed with separate PR

- [x] Replace `_.assign` with `Object.assign` (**Addressed with #7766**)
- [ ] Refactor `_.chain` to flow constructed with native functions
- [ ] Replace `_.compact` with `array.filter(Boolean)`
- [ ] Replace `_.concat` with `array.concat`
- [x] Replace `_.each(Array)` with `array.forEach` for array cases (**Addressed with #7748**)
- [ ] Replace `_.each(Object)` with `Object.entries().foreach` once we have #7362
- [x] Replace `_.forEach(Array)` with `array.forEach` for array cases (**Addressed with #7748**)
- [ ] Replace `_.forEach(Object)` with `Object.entries().foreach` once we have #7362
- [ ] Replace `_.entries` with `Object.entries` (can be done only after we have #7631)
- [ ] Replace `_.entriesIn` with `Object.entries` (can be done only after we have #7631). Note both are not equivalent, confirm that conversion is safe.
- [x] Replace `_.extend` with `Object.assign` Note they're not equivalent, but most likely `Object.assign` can be used safely in most of the cases (**Addressed with #7766**)
- [ ] Refactor `_.chain` to flow constructed with native functions
- [x] Replace `_.every` with `array.every` (**Addressed with  #7764**)
- [x] Replace `_.filter` with `array.filter`  (**Addressed with #7775**)
- [x] Replace `_.find` with `array.find`  (**Addressed with #7782**)
- [ ] Replace `_.findKey` with `Object.keys(object).find`
- [x] Replace `_.first` with `array[0]`  (**Addressed with #7816**)
- [ ] Replace `_.forOwn` with `Object.entries().forEach` (can be done only after we have #7362)
- [ ] Replace `_.get` with normal `obj.prop` checks, in top level property checks, or where we're safe from `Undefined is not an object` errors
- [ ] Replace `_.has` with normal `obj.prop` checks, in top level property checks, or where we're safe from `Undefined is not an object` errors
- [x] Replace `_.head` with `array[0]` (**Addressed with #7817**)
- [x] Replace `_.includes` with either `string.includes` or `array.includes` (**Addressed with #7818**)
- [ ] Replace `_.indexOf` with either `string.includes`, `array.includes` or `string.indexOf`, `array.indexOf`
- [ ] Remove `_.isBoolean`, we should just check whether value is _truthy_ or _falsy_, no need to confirm on _boolean_ type
- [ ] Replace `_.isEmpty` with `array.length` (only when it's applied on arrays)
- [ ] Replace `_.isEqual` with `===`
- [ ] Remove `_.isInteger`, in most cases it's used, validation is superfluous
- [x] Replace `_.isFunction` with `typeof value === 'function'` (**Addressed with #7810**)
- [x] Replace `_.isNil` with `value == null` (**Addressed with #7809**)
- [x] Replace `_.isString` with `typeof value === 'string'` (**Addressed with #7812**)
- [ ] Replace `_.isUndefined` with `value == null`
- [x] Replace `_.join` with `array.join` (**Addressed with #7805**)
- [ ] Replace `_.keyBy(collection)` with simple  _set_ collection
- [x] Replace `_.keys` with `Object.keys`  (**Addressed with #7784**)
- [ ] Replace `_.map` with `array.map`
- [ ] Replace `_merge` with `Object.assign`. Note they're not equivalent  (`_.merge` does deep merge) still `_.merge` is overused, and in most of the contexts `Object.assign` can be safely ised instead
- [ ] Replace `_.min` with `Math.min(...array)`
- [ ] Replace `_.nth` with `arr[index]`
- [ ] Replace `_.parseInt` with `Number`
- [ ] Replace `_.pullAllWith` with `array.filter`
- [ ] Replace `_.reduce` with `array.reduce` (in this case I'd also investigate if `reduce` is right pattern used)
- [ ] Replace `_.repeat` with `string.repeat`
- [ ] Replace `_.replace` with `string.replace`
- [ ] Replace `_.set` with normal `obj.prop = value` assignments, in top level property checks, or where we're safe from `Undefined is not an object` errors
- [x] Replace `_.size` with `array.length` check on arrays and `Object.keys(object).length on objects  (**Addressed with #7798**)
- [ ] Replace `_.some` with `array.some`
- [ ] Replace `_.sortBy` with `array.sort`
- [ ] Replace `_.split` with `string.split`
- [ ] Replace `_.takeRight` with `array.slice`
- [ ] Replace `_.toPairs` with `Object.entries` (can be done only after we have #7631)
- [ ] Replace `_.toString` with `String(value)` Note: behavior is not equivalent for `null` and `undefined`. We need to be sure to handle cases properly
- [x] Replace `_.toUpper` with `string.toUpper` (**Addressed with #7808**)
- [ ] Replace `_.unset` with `delete obj.prop`
- [ ] Replace `_.values` with `Object.values` (can be done only after we have #7631)

**Comment by medikoo** [comment]
@exoego great thanks for putting that into issue.

I've updated description with list of all functions which I believe we can replace. I think best if each is handled with separate PR's, so we process it in small steps, and change touches sensitive parts.

I've also added `good first issue` and `help wanted` labels as it appears there's a lot to cover

**Comment by chrisVillanueva** [comment]
Some news about lodash.  There's a security issue noted in this issue:

https://github.com/lodash/lodash/issues/4738

The current  version fails 3rd party security scans.

**Comment by chrisVillanueva** [comment]
@medikoo 
Is there a current file or sub-directory we can place utility functions?  Where can we add a function to execute deep and/or strict equality Object checks? I looked at `/utils` and `/lib/classes/Utils.js`.  I didn't see a place to introduce a general, shareable utility function.  `Utils.js` defines a class, but I only need a simple function to use across files. Any suggestions?

**Comment by medikoo** [comment]
> Is there a current file or sub-directory we can place utility functions?

@chrisVillanueva if you mean a language level utilities, then I think we should not introduce any here, but instead rely of utility library of choice. Currently we rely on lodash, and problem is that we overuse it for cases which should rather be handled with native API, and this issue is purely about that.

We may at some point consider replacing lodash completely, but I think then it should replaced with other utility library. We definitely should not attempt to build and maintain lang utilities in scope of this project.

**Comment by chrisVillanueva** [comment]
Thanks for clarification. This simplifies the issue.

**Comment by jishnu-mohan** [comment]
@exoego can we add`_.isObject` into this list?

**Comment by medikoo** [comment]
> @exoego can we add_.isObject into this list?

@jishnu-mohan Thanks for suggestion, still that one will have to be replaced with:

```javascript
value && (typeof value === 'function' || typeof value === 'object')
```

which is not readable. I'd say let's keep `_.isObject` now

**Comment by vietduc01100001** [comment]
@medikoo Can we add `_.flatten` to this list? As we discussed in #7862, we still need to keep `_.flatten` to ensure Node 6 compatibility. I found some ways we can achieve the same function with native JS [here](https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_flatten), basically making use of `Array.reduce` and `Array.concat`.

**Comment by medikoo** [comment]
>  I found some ways we can achieve the same function with native JS here, basically making use of Array.reduce and Array.concat.

@vietduc01100001 thanks for asking. Still I would leave `_.flatten` until we're able to replace it with native `[].flat`, and that'll be the case once we drop support for Node.js v10 (which we can consider next year in May). 

I didn't add it to the list, as it seems too early to speak about that.

**Comment by pgrzesik** [comment]
Hello, are the ones that are left up for grabs? I would like to pick up replacing `_.some` usage :+1:

**Comment by medikoo** [comment]
@pgrzesik yes, all unchecked are to take. Go for it!

**Comment by andreizet** [comment]
Hey, I'm going to replace ```_.has```, is that ok?

**Comment by medikoo** [comment]
> Hey, I'm going to replace _.has, is that ok?

@andreizet definitely, go for it!

**Comment by peterchou3788** [comment]
Hello, I'd like to work on _.get if its not taken.

**Comment by medikoo** [comment]
@peterchou3788 please go for it! Thank you

**Comment by peterchou3788** [comment]
@medikoo Is there a place to ask questions like slack or is it fine to ask them here?

**Comment by medikoo** [comment]
> @medikoo Is there a place to ask questions like slack or is it fine to ask them here?

If it's related to this issue, it's fine to ask here. It's unrelated general support matter please use slack or our forum

**Comment by peterchou3788** [comment]
Thank you. I noticed the variable `cfTemplate` in the file PickWebsocketsTemplatePart.js and it holds `resource` information. If I'm understanding this right, resources is either an object or dictionary but I couldn't determine the actual type.

**Comment by medikoo** [comment]
> resources is either an object or dictionary but I couldn't determine the actual type.

In JS there's just object type (no dictionary, it's just that object conceptually can be used as dictionary).

> in the file PickWebsocketsTemplatePart.js

As in this file through `_.get` we access eventual deep level property this one is not safe for refactor.

**Comment by medikoo** [comment]
As we dropped support for Node.js v8 (with https://github.com/serverless/serverless/issues/7362). It's now ok to pursue replacements that were in _on hold_ list!

I've put those items into main list in top description

**Comment by G-Rath** [comment]
Looks like `_.reduce` and `_.forOwn` are done, but not checked off :)

Also if you've got a build tool like TS or Babel, you could use optional chaining to replace most of the `_.get` usages (or something like `ts-optchain` to replace all the usages).

**Comment by medikoo** [comment]
> Looks like _.reduce and _.forOwn are done, but not checked off :)

@G-Rath thanks for pointing. Updated description 

> Also if you've got a build tool like TS or Babel, you could use optional chaining to replace most of the _.get usages (or something like ts-optchain to replace all the usages).

Once support for Node.js v12 is dropped we will simply move to native syntax. We do not plan to introduce any transpilation step into this package, as it comes with significant maintenance cost

**Comment by pjmattingly** [comment]
I'd like to work on:

> Replace \_.merge with Object.assign. Note they're not equivalent (_.merge does deep merge), so it should be done with care in all places where it's safe to rely on Object.assign instead (majority of cases)

If that would be okay. It looks like it's still up for grabs.

**Comment by pgrzesik** [comment]
Sure thing, that would be awesome @pjmattingly 👍

**Comment by medikoo** [comment]
I've added also `_.flat` and `_.flatMap` to the list, as those can be replaced now since we dropped support for Node.js v10

**Comment by pjmattingly** [comment]
I'd like to take `Replace _.flat with array.flat`

**Comment by pjmattingly** [comment]
Next, on to: `Replace _.flatMap with array.flatMap`

**Comment by pjmattingly** [comment]
Next, `Replace _.set with normal obj.prop = value assignments, but only in top level property checks or where given property path is ensured to exist`

**Comment by pjmattingly** [comment]
Then, the last one: `Replace _.get with normal obj.prop checks, but only in top level property checks or where given property path is ensured to exist`

**Comment by duytrandev** [comment]
hi @medikoo what about the case with `_.isObject(obj)`? I see there is a mix between lodash and using `const isObject = require('type/object/is')`

**Comment by medikoo** [comment]
> hi @medikoo what about the case with _.isObject(obj)? I see there is a mix between lodash and using const isObject = require('type/object/is')

@duytrandev I'd welcome PR that switches to `type/object/is`, still it's that's not really a scope here. Point of this issue is get rid of lodash in cases where native counterpart can be used, and we don't have plans to get rid of it completely (at least not at this point)
